### PR TITLE
Store: Get coupons via WooCommerce API

### PR DESCRIPTION
**Pull request by coderkevin**
This adds the data-layer http code necessary to fetch pages of coupons
from the API.

To Test:
1. `npm test`
2. `npm start`
3. Visit `http://calypso.localhost:3000/store/promotions/<site url>`
4. Open Redux Dev Tools and filter actions by `COUPONS`
5. View the `WOOCOMMERCE_COUPONS_PAGE_UPDATED` action and verify data in the fields matches what you expect from your site.

### CODE REVIEW

File: client/extensions/woocommerce/state/sites/coupons/actions.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+	WOOCOMMERCE_COUPONS_REQUEST,
+	WOOCOMMERCE_COUPONS_PAGE_UPDATED,
+} from 'woocommerce/state/action-types';
+
+export function fetchCouponsPage( siteId, pageIndex = 1, perPage = 10 ) {
+	return { type: WOOCOMMERCE_COUPONS_REQUEST, siteId, pageIndex, perPage };
+}
+
+export function couponsPageUpdated( siteId, pageIndex, coupons, totalPages, totalCoupons ) {
+	return { type: WOOCOMMERCE_COUPONS_PAGE_UPDATED, siteId, pageIndex, coupons, totalPages, totalCoupons };
```
Comment (justinshreve) [inline client/extensions/woocommerce/state/sites/coupons/actions.js]: I think the general pattern we wanted to go with here was `*_REQUEST` and `*_RECEIVE` actions. I had a similar comment over [here](https://github.com/Automattic/wp-calypso/pull/17922#discussion_r137664494).

File: client/extensions/woocommerce/app/promotions/index.js [hunk 2]
```diff
@@ -24,6 +26,23 @@ class Promotions extends Component {
...
+	componentDidMount() {
+		const { site } = this.props;
+		if ( site && site.ID ) {
+			this.props.fetchCouponsPage( site.ID, 1 );
+		}
+	}
+
+	componentWillReceiveProps( newProps ) {
+		const { site } = this.props;
+		const newSiteId = newProps.site && newProps.site.ID || null;
```
Comment (timmyc) [inline client/extensions/woocommerce/app/promotions/index.js]: This and the next line could be simplified using `_.get` but fine as-is too.

File: client/extensions/woocommerce/state/sites/coupons/actions.js (hunk 1 above)
Comment (coderkevin) [reply client/extensions/woocommerce/state/sites/coupons/actions.js]: `_RECEIVE` actions aren't so desirable anymore, as we push for further separation between the code which needs data, and the code which does the actual fetching. The preference now is for more meaningful type names. That's why I chose this type name here.

File: client/extensions/woocommerce/app/promotions/index.js (hunk 2 above)
Comment (coderkevin) [reply client/extensions/woocommerce/app/promotions/index.js]: Yeah, I didn't figure it was worth adding another import just for one line.

**Comment by coderkevin** [comment]
Changed `WOOCOMMERCE_COUPONS_REQUEST` to `WOOCOMMERCE_COUPONS_REQUEST_PAGE` for clarification.

File: client/extensions/woocommerce/state/sites/coupons/test/handlers.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+ * Internal dependencies
+ */
+import { WPCOM_HTTP_REQUEST } from 'state/action-types';
+import { fetchCouponsPage, couponsPageUpdated } from '../actions';
+import {
+	requestCouponsPage,
+	requestCouponsPageSuccess
+} from '../handlers';
+
+describe( 'handlers', () => {
```
Comment (allendav) [inline client/extensions/woocommerce/state/sites/coupons/test/handlers.js]: Please consider adding a test to ensure that the coupons reducers apply changes only to the appropriate site.

Comment (coderkevin) [reply client/extensions/woocommerce/state/sites/coupons/test/handlers.js]: The reducer for this is in #18148. Is that what you meant?

**Comment by coderkevin** [comment]
Rebased.

**Pull request merged**